scipy/special/_basic.py

```diff
@@ -2911,34 +2911,66 @@ def _factorialx_array_exact(n, k=1):
     return out
 
 
-def _factorialx_array_approx(n, k):
+def _factorialx_array_approx(n, k, extend):
     """
     Calculate approximation to multifactorial for array n and integer k.
 
-    Ensure we only call _factorialx_approx_core where necessary/required.
+    Ensure that values aren't calculated unnecessarily.
     """
+    if extend == "complex":
+        return _factorialx_approx_core(n, k=k, extend=extend)
+
+    # at this point we are guaranteed that extend='zero' and that k>0 is an integer
     result = zeros(n.shape)
     # keep nans as nans
     place(result, np.isnan(n), np.nan)
     # only compute where n >= 0 (excludes nans), everything else is 0
     cond = (n >= 0)
     n_to_compute = extract(cond, n)
-    place(result, cond, _factorialx_approx_core(n_to_compute, k=k))
+    place(result, cond, _factorialx_approx_core(n_to_compute, k=k, extend=extend))
     return result
 
 
-def _factorialx_approx_core(n, k):
+def _factorialx_approx_core(n, k, extend):
     """
     Core approximation to multifactorial for array n and integer k.
     """
+    def _gamma_no_inf(vals):
+        res = gamma(vals)
+        # replace infinities coming from gamma function with nan, see #19071;
+        # gamma only returns inf for real inputs, so no fix needed for complex inputs
+        if isinstance(res, np.ndarray):
+            if not _is_subdtype(vals.dtype, "c"):
+                # we _do_ keep infinity where input itself was +∞
+                keep_as_inf = np.isinf(vals) & (vals > 0)
+                res[np.isinf(res) & ~keep_as_inf] = np.nan
+        elif np.isinf(res) and not (np.isinf(vals) and (vals > 0)):
+            res = np.float64("nan")
+        return res
```

Comment:
@h-vetinari, at this point, I think we should just fix `gamma` and related functions. Hopefully there aren't actually too many users relying on this bug in the behavior at the poles, but if so, it should be fine to just add a note in the docstring explaining the old and new behavior, and to use `rgamma` instead of dividing by the `gamma` function.

I'll make an issue and discourse post and see if we can get the fix through this week.

Sorry, I was the one dragging my feet on fixing this for fear of breaking someones workflow, but seeing the workarounds you had to do has convinced me.

I'll make an issue about it and see if we can get this through

Reply:
I had completely forgotten about the discussion in #19071 already - that would probably be a good candidate for the issue you're talking about? Happy to switch to `rgamma` of course.

Reply:
Fixed as discussed in #21827